File: samples/RdProtocol/src/dotnet/ReSharperPlugin.RdProtocol/Rider/Component.cs
using System;
using System.Threading.Tasks;
using JetBrains.Lifetimes;
using JetBrains.ProjectModel;
using JetBrains.Rd.Tasks;
using JetBrains.RdBackend.Common.Features;
using ReSharperPlugin.RdProtocol;

namespace ReSharperPlugin.TestPlugin
{
    [SolutionComponent]
    public class Component
    {
        public Component(Lifetime lifetime, ISolution solution)
        {
            // Get the model
            var model = solution.GetProtocolSolution().GetRdProtocolModel();

            // Properties
            model.Property.Value = new CustomType(
                @string: "value",
                boolean: true,
                array: new[] { "first", "second" });

            model.Property.Advise(lifetime, x =>
            {
                // Handle changes
            });

            // Maps
            model.Map["key"] = "value";
            var mapValue = model.Map["key"];

            model.Map.Advise(lifetime, x =>
            {
                // Handle changes
                var mapKey = x.Key;
            });

            // Call frontend remote procedure
            model.Callback.Sync("args");

            // Set backend remote procedure
            model.Call.Set((lt, x) => Task.FromResult(new[] { x }).ToRdTask());

            // Trigger backend event
            model.Sink("value");

            // Subscribe frontend event
            model.Source.Advise(lifetime, x => { });
        
            // Trigger/subscribe bidirectional event
            model.Signal.Fire("value");
            model.Signal.Advise(lifetime, x => { });
        }
    }
}

File: samples/RdProtocol/src/rider/main/kotlin/com/jetbrains/rider/plugins/rdprotocol/Component.kt
package com.jetbrains.rider.plugins.rdprotocol

import com.intellij.openapi.project.Project
import com.jetbrains.rdclient.util.idea.LifetimedProjectComponent
import com.jetbrains.rider.projectView.solution

class Component(project: Project) : LifetimedProjectComponent(project) {
    init {
        // Get the model
        val model = project.solution.rdProtocolModel

        // Properties
        model.property.set(
            CustomType(
                string = "value",
                boolean = true,
                array = arrayOf("first", "second")
            )
        )

        model.property.advise(componentLifetime) {
            // Handle changes
        }

        // Maps
        model.map["key"] = "value"
        val mapValue = model.map["key"]

        model.map.advise(componentLifetime) {
            // Handle changes
            var mapKey = it.key
        }

        // Call backend remote procedure
        model.call.sync("args")

        // Set frontend remote procedure
        model.callback.set { arg -> arrayOf(arg) }

        // Trigger frontend event
        model.source.fire("value")

        // Subscribe backend event
        model.sink.advise(componentLifetime) { }

        // Trigger/subscribe bidirectional event
        model.signal.fire("value")
        model.signal.advise(componentLifetime) { }
    }
}
